common/setups/hybrid/gmm_system.py

```diff
@@ -646,210 +682,535 @@ def recognition(
                     **kwargs,
                 )
 
+    def sat_recognition(
+        self,
+        prev_ctm: str,
+        feature_cache: str,
+        cache_regex: str,
+        cmllr_mixtures: str,
+        train_corpus: str,
+        name: str,
+        iters: List[int],
+        lm_scales: List[float],
+        feature_scorer: Tuple[str, str],
+        optimize_am_lm_scale: bool,
+        corpus: str,
+        feature_flow: str,
+        pronunciation_scale: Union[float, List[float]],
+        search_parameters: dict,
+        rtf: float,
+        mem: float,
+        parallelize_conversion: bool,
+        lattice_to_ctm_kwargs: dict,
+        **kwargs,
+    ):
+        pronunciation_scale = (
+            [pronunciation_scale]
+            if isinstance(pronunciation_scale, float)
+            else pronunciation_scale
+        )
+
+        for it, p, l in itertools.product(iters, pronunciation_scale, lm_scales):
+            recognized_corpus = corpus_recipes.ReplaceTranscriptionFromCtmJob(
+                self.corpora[corpus].corpus_file,
+                self.ctm_files[corpus][
+                    f"recog_{train_corpus}-{prev_ctm}-{corpus}-ps{p:02.2f}-lm{l:02.2f}-iter{it:02d}"
+                ],
+            )
+            speaker_seq = corpus_recipes.SegmentCorpusBySpeakerJob(
+                self.corpora[corpus].corpus_file
+            )
+
+            overlay = "%s_sat" % corpus
+            self.add_overlay(corpus, overlay)
+            self.crp[overlay].corpus_config = copy.deepcopy(
+                self.crp[corpus].corpus_config
+            )
+            self.crp[overlay].corpus_config.file = recognized_corpus.output_corpus_path
+            self.crp[overlay].segment_path = copy.deepcopy(
+                self.crp[corpus].segment_path
+            )
+
+            self.corpora[overlay] = copy.deepcopy(self.corpora[corpus])
+            self.corpora[overlay].corpus_file = recognized_corpus.output_corpus_path
+
+            alignment = mm.AlignmentJob(
```

Comment:
at the moment I use one loop over `iters`, `pronunciation_scales`, and `lm_scales`. So move the complete `add_overlay` and `AlignmentJob` out of the loop?

Reply:
I have been re-reading the code and it I noticed I was wrong. While indeed the alignment would be Independent of the LM scale, the LM scale determines which transcription (i.e. ctm file of previous recognition run with LM_scale=l) is used for the alignment (cf. line 717). So currently the implementation is correct.

But now I wonder if the overlay you are creating is correct. Because for every `it, p, l` you are creating the same overlay `"%s_sat" % corpus` despite the fact that the corpus file is different. 
Probably the correct thing happens, but after the loop the older overlays are inaccessible/overwritten.

Reply:
I added the `it, p, l` to the `overlay_key`